Framework/Algorithms/src/PolarizationCorrections/PolarizerEfficiency.cpp

```diff
@@ -0,0 +1,188 @@
+  const auto &spinValidator = std::make_shared<SpinStateValidator>(std::unordered_set<int>{4});
+  declareProperty(PropertyNames::SPIN_STATES, "11,10,01,00", spinValidator,
+                  "Order of individual spin states in the input group workspace, e.g. \"01,11,00,10\"");
+
+  declareProperty(std::make_unique<FileProperty>(PropertyNames::OUTPUT_FILE_PATH, "", FileProperty::OptionalSave),
+                  "File name or path for the output to be saved to.");
+}
+
+/**
+ * Tests that the inputs are all valid
+ * @return A map containing the incorrect workspace
+ * properties and an error message
+ */
+std::map<std::string, std::string> PolarizerEfficiency::validateInputs() {
+  std::map<std::string, std::string> errorList;
+  const WorkspaceGroup_sptr inputWorkspace = getProperty(PropertyNames::INPUT_WORKSPACE);
+  if (inputWorkspace == nullptr) {
+    errorList[PropertyNames::INPUT_WORKSPACE] = "The input workspace is not a workspace group.";
+    return errorList;
+  }
+
+  if (inputWorkspace->size() != 4) {
+    errorList[PropertyNames::INPUT_WORKSPACE] =
+        "The input group workspace must have four periods corresponding to the four spin configurations.";
+  } else {
+    for (size_t i = 0; i < inputWorkspace->size(); ++i) {
+      const MatrixWorkspace_sptr stateWs = std::dynamic_pointer_cast<MatrixWorkspace>(inputWorkspace->getItem(i));
+      Unit_const_sptr unit = stateWs->getAxis(0)->unit();
+      if (unit->unitID() != "Wavelength") {
+        errorList[PropertyNames::INPUT_WORKSPACE] = "All input workspaces must be in units of Wavelength.";
+      }
+    }
+  }
+
+  // Check outputs.
+  auto const &outputWs = getPropertyValue(PropertyNames::OUTPUT_WORKSPACE);
+  auto const &outputFile = getPropertyValue(PropertyNames::OUTPUT_FILE_PATH);
+  if (outputWs.empty() && outputFile.empty()) {
+    errorList[PropertyNames::OUTPUT_FILE_PATH] = "Either an output workspace or output file must be provided.";
+    errorList[PropertyNames::OUTPUT_WORKSPACE] = "Either an output workspace or output file must be provided.";
+  }
+
+  return errorList;
+}
+
+/**
+ * Explicitly calls validateInputs and throws runtime error in case
+ * of issues in the input properties.
+ */
+void PolarizerEfficiency::validateGroupInput() {
+  const auto &results = validateInputs();
+  if (results.size() > 0) {
+    const auto &result = results.cbegin();
+    throw std::runtime_error("Issue in " + result->first + " property: " + result->second);
+  }
+}
+
+bool PolarizerEfficiency::processGroups() {
+  validateGroupInput();
+  calculatePolarizerEfficiency();
+  return true;
+}
+
+void PolarizerEfficiency::exec() { calculatePolarizerEfficiency(); }
+
+void PolarizerEfficiency::calculatePolarizerEfficiency() {
+  // First we extract the individual workspaces corresponding to each spin configuration from the group workspace
+  const auto &groupWorkspace =
+      AnalysisDataService::Instance().retrieveWS<WorkspaceGroup>(getProperty(PropertyNames::INPUT_WORKSPACE));
```

Comment:
On this, and relevant to the other comment about simplifying the validation, have we done some testing to see if you really can pass in a workspace that's not in the ADS to a `WorkspaceGroup` property? I did what I consider to be some very incomplete testing, but it seemed to show that the property type was enforcing that the workspace group had to be in the ADS. It would be good to establish this conclusively, as it will be relevant to all the other polarised SANS algorithms.

Reply:
Yes, I can confirm this. Below shows the debugging from C++ in the tests. The `CreateSampleWorkspace` outputs are present in the ADS due to it being a non-child alg, but the group is not as the `GroupWorkspaces` alg is a child algorithm. The group only exists as a pointer in the C++ code. It's then passed to the `FlipperEfficiency` alg without entering the ADS which is able to process normally. 
![image](https://github.com/mantidproject/mantid/assets/47181718/41bc272f-d1e3-43f0-9ee1-8fc109561601)

Reply:
OK, brilliant, thanks for checking 👍